tmpl/wiki/error-handling.md

```diff
@@ -0,0 +1,384 @@
+> **Having every operation marked as one that might fail is no more explicit than having none of them marked.**
+>
+> In short, for errors that are a foreseeable and ordinary part of the execution of your production code and that are not omnipresent, error-aware return types are typically the right solution.
+
+In the case of Mirage, it is typically the case that every operation may fail and that, since the interfaces are abstract, we cannot know all the ways they may fail.
+This seems like a good argument for using exceptions.
+
+However, we should still use explicit variants for expected (non-exceptional) cases that callers will probably want to handle.
+For example, `KV_RO`'s `Unknown_key` case is something code will often want to use, e.g. to provide a default value.
+This probably shouldn't even be considered an error.
+
+There are several systems we could use:
+
+1. Interfaces define concrete error types as polymorphic variants (the current system).
+2. Interfaces define abstract error types, plus functions to convert to a `string` or `sexp`.
+3. All interfaces use the single general-purpose `Error.t` type.
+4. Interfaces throw exceptions.
+
+System 2 (using abstract types) has the advantage that we can support this without breaking existing APIs (updated code still satisfies the old API):
+
+```
+module type BLOCK = sig
+  type error with sexp
+  val error_message : error -> string
+
+  val read:
+    t -> int64 -> page_aligned_buffer list ->
+    [ `Error of error | `Ok of unit ] io
+```
+
+System 3 allows callers to handle errors generically, while the `Or_error.t` provides functions to convert to a `string` or `Sexp.t` automatically:
+
+```
+module type BLOCK = sig
+  val read:
+    t -> int64 -> page_aligned_buffer list ->
+    unit Or_error.t io
+```
+
+System 4 removes all mention of errors from the interface, while `sexp_of_exn` and `Printexc.to_string` still provide `Sexp.t` and `string` values:
+
+```
+module type BLOCK = sig
+  val read:
+    t -> int64 -> page_aligned_buffer list ->
+    unit io
+```
+
+It may be useful to predefine some common errors (e.g. `Connection_refused`) as exceptions in `mirage-types`.
+
+Looking back at the three uses above:
+
+- For providing human readable errors, any system except 1 will do.
+- Allowing programs to detect and handle certain errors specially works best with systems 1 and 4.
+  System 2 only works if callers require a particular concrete implementation, breaking the abstraction.
+  System 3 provides (I think) no easy way to match on the type of error.
+- For signalling errors in order to trigger rollback or freeing of resources, any system will do.
+
+It is perhaps worth noting that handling specific error cases specially is extremely rare.
+I don't think I've seen any examples of this feature being used with the `Error` variants in the Mirage code-base.
```

Comment:
- If the transmit buffer is full, we just wait until there's space.
- I'd say `EAGAIN` is a perfect example of something that should *not* be an exception. It's something you probably want to handle, and will probably forget unless prompted. Let's not have any in our APIs ;-)
- I'd be tempted to say that connection refused should be a return code too (`Network_error of exn`), since unreliable networks are something else you might often want to handle. The payload should be an exception so people who don't care can just throw it.